dds/DCPS/DiscoveryBase.h

```diff
@@ -892,6 +913,54 @@ namespace OpenDDS {
         return -1;
       }
 
+
+      struct MatchingData {
+        RepoId writer;
+        RepoId reader;
+        SequenceNumber rpc_sequence_number;
+        const DDS::DataWriterQos* dwQos;
+        const DDS::PublisherQos* pubQos;
+        const DDS::DataReaderQos* drQos;
+        const DDS::SubscriberQos* subQos;
+        TransportLocatorSeq* wTls;
+        TransportLocatorSeq* rTls;
+        XTypes::TypeInformation* reader_type_info;
+        XTypes::TypeInformation* writer_type_info;
+        const ContentFilterProperty_t* cfProp;
+        MonotonicTimePoint time_added_to_map;
+      };
+
+      struct MatchingPair {
```

Comment:
std::pair does not seem to have 'less than' operator. We can derive our matching pair class from std::pair and add 'less than' operator, but I am not sure it would be much better, than what we have (?).

Reply:
https://en.cppreference.com/w/cpp/utility/pair/operator_cmp

Reply:
Interesting... When I tried it yesterday, compiler complained and when I looked at the class source I did not see it there. I have implemented it now and it compiles.

Reply:
I misspoke - it still does not compile

Reply:
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\utility(265,1): error C2678:  binary '<': no operator found which takes a left-hand operand of type 'const _Ty1' (or there is no acceptable conversion)
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\utility(265,1): error C2678:         with
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\utility(265,1): error C2678:         [
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\utility(265,1): error C2678:             _Ty1=OpenDDS::DCPS::RepoId
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\utility(265,1): error C2678:         ]
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\system_error(263,24): message :  could be 'bool std::operator <(const std::error_condition &,const std::error_condition &) noexcept'
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\system_error(258,24): message :  or       'bool std::operator <(const std::error_code &,const std::error_code &) noexcept'
1>C:\Users\kuznetsovm\source\repos\OpenDDSX\dds\DCPS\TimeDuration.h(83,26): message :  or       'bool OpenDDS::DCPS::operator <(const OpenDDS::DCPS::TimeDuration &,const OpenDDS::DCPS::TimeDuration &)' [found using argument-dependent lookup]
1>C:\Users\kuznetsovm\source\repos\OpenDDSX\dds\DCPS\Time_Helper.h(61,6): message :  or       'bool OpenDDS::DCPS::operator <(const DDS::Duration_t &,const DDS::Duration_t &)' [found using argument-dependent lookup]
1>C:\Users\kuznetsovm\source\repos\OpenDDSX\dds\DCPS\Time_Helper.h(84,6): message :  or       'bool OpenDDS::DCPS::operator <(const DDS::Time_t &,const DDS::Time_t &)' [found using argument-dependent lookup]
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\utility(265,1): message :  while trying to match the argument list '(const _Ty1, const _Ty1)'
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\utility(265,1): message :         with
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\utility(265,1): message :         [
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\utility(265,1): message :             _Ty1=OpenDDS::DCPS::RepoId
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.22.27905\include\utility(265,1): message :         ]

Reply:
Right, it would delegate to `operator<` for the GUID which would need to be provided just like the Duration_t/Time_t ones that the compiler is pointing out.

This is not essential, so if it becomes more trouble than it's worth we can go with the code as-is.  However I'd like to rely on the standard library types as much as we can.